lib/Bridge/TolerantParser/WorseReflection/WorseSignatureHelper.php

```diff
@@ -0,0 +1,171 @@
+        }
+
+        if ($callable instanceof MemberAccessExpression) {
+            $reflectionOffset = $this->reflector->reflectOffset($textDocument, $offset->toInt());
+            $symbolContext = $reflectionOffset->symbolContext();
+
+            if ($symbolContext->symbol()->symbolType() !== Symbol::METHOD) {
+                throw new CouldNotHelpWithSignature(sprintf(
+                    'Could not provide signature member type "%s"',
+                    $symbolContext->symbol()->symbolType()
+                ));
+            }
+
+            $containerType = $symbolContext->containerType();
+
+            if (null === $containerType) {
+                throw new CouldNotHelpWithSignature(sprintf(
+                    'Could not determine container type for: "%s"',
+                    $symbolContext->symbol()->name()
+                ));
+            }
+
+            if (!$containerType->isClass()) {
+                throw new CouldNotHelpWithSignature(sprintf(
+                    'Container type is not a class: "%s"',
+                    $symbolContext->symbol()->name()
+                ));
+            }
+
+            $reflectionClass = $this->reflector->reflectClass($containerType->className());
+            $reflectionMethod = $reflectionClass->methods()->get($symbolContext->symbol()->name());
+
+            return $this->createSignatureHelp($reflectionMethod, $position);
+        }
+
+        throw new CouldNotHelpWithSignature(sprintf('Could not provide signature for AST node of type "%s"', get_class($callable)));
+    }
+
+    private function signatureHelpForFunction(QualifiedName $callable, int $position): SignatureHelp
+    {
+        $name = $callable->__toString();
+        $functionReflection = $this->reflector->reflectFunction($name);
+        
+        return $this->createSignatureHelp($functionReflection, $position);
+    }
+
+    private function createSignatureHelp(ReflectionFunctionLike $functionReflection, int $position)
+    {
+        $signatures = [];
+        $parameters = [];
+        
+        /** @var ReflectionParameter $parameter */
+        foreach ($functionReflection->parameters() as $parameter) {
+            $formatted = $this->formatter->format($parameter);
+            $parameters[] = new ParameterInformation($parameter->name(), $formatted);
+        }
+        
+        $formatted = $this->formatter->format($functionReflection);
+        $signatures[] = new SignatureInformation($formatted, $parameters);
+        
+        return new SignatureHelp($signatures, $position);
+    }
+
+    private function signatureHelpForScopedPropertyAccess(ScopedPropertyAccessExpression $callable, CallExpression $node, int $position)
+    {
+        $scopeResolutionQualifier = $callable->scopeResolutionQualifier;
+        
+        if (!$scopeResolutionQualifier instanceof QualifiedName) {
+            throw new CouldNotHelpWithSignature(sprintf('Static calls only supported with qualified names'));
```

Comment:
Note: much of this logic is duplicated, should extract to helpers (and improve horizontally)